coderd/workspaceagentsrpc.go

```diff
@@ -184,73 +166,137 @@ func ensureLatestBuild(ctx context.Context, db database.Store, logger slog.Logge
+	conn *websocket.Conn,
+) {
+	minder := &agentWebsocketMinder{
+		apiCtx:            api.ctx,
+		workspaceAgent:    workspaceAgent,
+		workspaceBuild:    workspaceBuild,
+		conn:              conn,
+		pingPeriod:        api.AgentConnectionUpdateFrequency,
+		db:                api.Database,
+		replicaID:         api.ID,
+		updater:           api,
+		disconnectTimeout: api.AgentInactiveDisconnectTimeout,
+		logger:            api.Logger.With(slog.F("workspace_id", workspaceBuild.WorkspaceID)),
+	}
+	minder.init()
+	go pprof.Do(ctx, pprof.Labels("agent", workspaceAgent.ID.String()), minder.sendPings)
+	go pprof.Do(ctx, pprof.Labels("agent", workspaceAgent.ID.String()), minder.mind)
+}
+
+type workspaceUpdater interface {
+	publishWorkspaceUpdate(ctx context.Context, workspaceID uuid.UUID)
+}
+
+type pingerCloser interface {
+	Ping(ctx context.Context) error
+	Close(code websocket.StatusCode, reason string) error
+}
+
+type agentWebsocketMinder struct {
+	apiCtx         context.Context
+	workspaceAgent database.WorkspaceAgent
+	workspaceBuild database.WorkspaceBuild
+	conn           pingerCloser
+	db             database.Store
+	replicaID      uuid.UUID
+	updater        workspaceUpdater
+	logger         slog.Logger
+	pingPeriod     time.Duration
+
+	// state manipulated by both sendPings() and mind() goroutines: needs to be threadsafe
+	lastPing atomic.Pointer[time.Time]
+
+	// state manipulated only by mind() goroutine: does not need to be threadsafe
+	firstConnectedAt  sql.NullTime
+	lastConnectedAt   sql.NullTime
+	disconnectedAt    sql.NullTime
+	disconnectTimeout time.Duration
+}
+
+// sendPings sends websocket pings.
+//
+// We use a custom heartbeat routine here instead of `httpapi.Heartbeat`
+// because we want to log the agent's last ping time.
+func (m *agentWebsocketMinder) sendPings(ctx context.Context) {
+	t := time.NewTicker(m.pingPeriod)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-t.C:
+		case <-ctx.Done():
+			return
+		}
+
+		// We don't need a context that times out here because the ping will
+		// eventually go through. If the context times out, then other
+		// websocket read operations will receive an error, obfuscating the
+		// actual problem.
+		err := m.conn.Ping(ctx)
+		if err != nil {
+			return
 		}
+		m.lastPing.Store(ptr.Ref(time.Now()))
+	}
+}
+
+func (m *agentWebsocketMinder) updateConnectionTimes(ctx context.Context) error {
+	//nolint:gocritic // We only update the agent we are minding.
+	err := m.db.UpdateWorkspaceAgentConnectionByID(dbauthz.AsSystemRestricted(ctx), database.UpdateWorkspaceAgentConnectionByIDParams{
+		ID:               m.workspaceAgent.ID,
+		FirstConnectedAt: m.firstConnectedAt,
+		LastConnectedAt:  m.lastConnectedAt,
+		DisconnectedAt:   m.disconnectedAt,
+		UpdatedAt:        dbtime.Now(),
+		LastConnectedReplicaID: uuid.NullUUID{
+			UUID:  m.replicaID,
+			Valid: true,
+		},
 	})
+	if err != nil {
+		return err
```

Comment:
Would be nice to wrap errors so we have a stack trace and better understanding where it came from.